pyro/poutine/indep_poutine.py

```diff
@@ -2,40 +2,34 @@
 
 from collections import namedtuple
 
-from .poutine import Poutine
+from .poutine import Messenger
 
 CondIndepStackFrame = namedtuple("CondIndepStackFrame", ["name", "counter", "vectorized"])
 
 
-class IndepPoutine(Poutine):
+class IndepMessenger(Messenger):
     """
-    This poutine keeps track of stack of independence information declared by
+    This messenger keeps track of stack of independence information declared by
     nested ``irange`` and ``iarange`` contexts. This information is stored in
     a ``cond_indep_stack`` at each sample/observe site for consumption by
     ``TracePoutine``.
     """
-    def __init__(self, fn, name, vectorized):
+    def __init__(self, name, vectorized):
         """
         Constructor: basically default, but store a counter to keep track of
         which ``irange`` branch we're in.
         """
+        super(IndepMessenger, self).__init__()
         self.name = name
         self.counter = 0
         self.vectorized = vectorized
-        super(IndepPoutine, self).__init__(fn)
 
-    def __enter__(self):
+    def next_context(self):
         """
-        Increment counter by one each time we enter a new ``irange`` branch.
+        Increments the counter.
         """
         self.counter += 1
-        return super(IndepPoutine, self).__enter__()
 
-    def _prepare_site(self, msg):
-        """
-        Construct the message that is consumed by ``TracePoutine``;
-        ``cond_indep_stack`` encodes the nested sequence of ``irange`` branches
-        that the site at name is within.
-        """
-        msg["cond_indep_stack"].append(CondIndepStackFrame(self.name, self.counter, self.vectorized))
-        return msg
+    def _process_message(self, msg):
+        msg["cond_indep_stack"].insert(0, CondIndepStackFrame(self.name, self.counter, self.vectorized))
```

Comment:
This erroneously reverses the direction of `cond_indep_stack`. It worries me that this bug was not caught by tests.

Reply:
cc @martinjankowiak

Reply:
> it reverses both the order of traversal of the poutine stack and the order of insertion into the
independence stack

Oh, that makes sense, thanks for explaining.

Reply:
I actually first just reversed the order of the poutine stack traversal and then your tests caught the problem with the independence stack order so we reversed that too.